// openclaw/dist/audio-preflight-LWdrCKEr.js
import "./paths-Bp5uKvNR.js";
import { J as logVerbose, Z as shouldLogVerbose } from "./subsystem-DNjXI6xf.js";
import "./agent-scope-DcKzlxc-.js";
import "./model-selection-C8bqH1yt.js";
import "./github-copilot-token-Bpxb7S7I.js";
import "./env-CtORg1Fr.js";
import "./plugins-CMqWDObz.js";
import "./accounts-ETt7H2v2.js";
import "./bindings-Cs1_iD7i.js";
import "./accounts-0krJczrZ.js";
import "./image-ops-D3E0ptta.js";
import "./pi-model-discovery-j5tVLINv.js";
import "./message-channel-EuGuj6kr.js";
import "./pi-embedded-helpers-CNhhELVT.js";
import "./config-Y4aFpi2A.js";
import "./manifest-registry-C6UHLS4e.js";
import "./dock-Sh2AiVEn.js";
import "./chrome-Cp11xsGp.js";
import "./ssrf-D_av5wos.js";
import "./skills-Deq7jLXB.js";
import "./redact-DSLz_jVb.js";
import "./errors-Ch4QboNP.js";
import "./store-DCsfOq9h.js";
import "./sessions-BH9VzbnR.js";
import "./accounts-DTsBe9PV.js";
import "./paths-BzZzj2Uc.js";
import "./tool-images-DqeXEfyj.js";
import "./thinking-CLSwmd68.js";
import "./image-_0vi_hl_.js";
import "./gemini-auth-D07WQyiR.js";
import "./fetch-guard-CSuEAlsd.js";
import "./local-roots-BzJ5LI6h.js";
import { a as resolveMediaAttachmentLocalRoots, n as createMediaAttachmentCache, o as runCapability, r as normalizeMediaAttachments, t as buildProviderRegistry, u as isAudioAttachment } from "./runner-cOolKiXH.js";

//#region src/media-understanding/audio-preflight.ts
/**
* Transcribes the first audio attachment BEFORE mention checking.
* This allows voice notes to be processed in group chats with requireMention: true.
* Returns the transcript or undefined if transcription fails or no audio is found.
*/
async function transcribeFirstAudio(params) {
	const { ctx, cfg } = params;
	const audioConfig = cfg.tools?.media?.audio;
	if (!audioConfig || audioConfig.enabled === false) return;
	const attachments = normalizeMediaAttachments(ctx);
	if (!attachments || attachments.length === 0) return;
	const firstAudio = attachments.find((att) => att && isAudioAttachment(att) && !att.alreadyTranscribed);
	if (!firstAudio) return;
	if (shouldLogVerbose()) logVerbose(`audio-preflight: transcribing attachment ${firstAudio.index} for mention check`);
	const providerRegistry = buildProviderRegistry(params.providers);
	const cache = createMediaAttachmentCache(attachments, { localPathRoots: resolveMediaAttachmentLocalRoots({
		cfg,
		ctx
	}) });
	try {
		const result = await runCapability({
			capability: "audio",
			cfg,
			ctx,
			attachments: cache,
			media: attachments,
			agentDir: params.agentDir,
			providerRegistry,
			config: audioConfig,
			activeModel: params.activeModel
		});
		if (!result || result.outputs.length === 0) return;
		const audioOutput = result.outputs.find((output) => output.kind === "audio.transcription");
		if (!audioOutput || !audioOutput.text) return;
		firstAudio.alreadyTranscribed = true;
		if (shouldLogVerbose()) logVerbose(`audio-preflight: transcribed ${audioOutput.text.length} chars from attachment ${firstAudio.index}`);
		return audioOutput.text;
	} catch (err) {
		if (shouldLogVerbose()) logVerbose(`audio-preflight: transcription failed: ${String(err)}`);
		return;
	} finally {
		await cache.cleanup();
	}
}

//#endregion
export { transcribeFirstAudio };
