// openclaw/dist/audio-preflight-B4kNpgjV.js
import "./paths-CyR9Pa1R.js";
import { J as logVerbose, Z as shouldLogVerbose } from "./subsystem-NZzGD1xL.js";
import "./agent-scope-DGv_lb5V.js";
import "./workspace-DSaPuILJ.js";
import "./plugins-DRxcIsdW.js";
import "./accounts-BdsaOk1K.js";
import "./boolean-mcn6kL0s.js";
import "./command-format-DkQnibeb.js";
import "./bindings-BAjLQTjN.js";
import "./accounts-Bnm1qsMy.js";
import "./image-ops-D_0HqKCn.js";
import "./model-auth-D7yExN0T.js";
import "./github-copilot-token-Dgb9dAHW.js";
import "./pi-model-discovery-DaNAekda.js";
import "./message-channel-CGfDQO-K.js";
import "./pi-embedded-helpers-DxTyisc4.js";
import "./config-Co-5D6Gn.js";
import "./manifest-registry-CBHbnE_M.js";
import "./dock-EZ2jebb_.js";
import "./chrome-B1nMBMAN.js";
import "./ssrf-DoQDLHSj.js";
import "./frontmatter-CGOZnTsi.js";
import "./skills-CfaId93d.js";
import "./redact-CpDh05JC.js";
import "./errors-xxnHvoWf.js";
import "./store-LAmaIatB.js";
import "./sessions-The2KDBw.js";
import "./accounts-CU1yRDCa.js";
import "./paths-DZd1zVsd.js";
import "./tool-images-Dq_CJUOG.js";
import "./thinking-Ds6fxj9u.js";
import "./image-Cl4RgYGS.js";
import "./gemini-auth-BkfXMw6M.js";
import "./fetch-guard-BlZG4Lmu.js";
import "./local-roots-BpEzzsZA.js";
import { a as resolveMediaAttachmentLocalRoots, n as createMediaAttachmentCache, o as runCapability, r as normalizeMediaAttachments, t as buildProviderRegistry, u as isAudioAttachment } from "./runner-CjFAR0P3.js";

//#region src/media-understanding/audio-preflight.ts
/**
* Transcribes the first audio attachment BEFORE mention checking.
* This allows voice notes to be processed in group chats with requireMention: true.
* Returns the transcript or undefined if transcription fails or no audio is found.
*/
async function transcribeFirstAudio(params) {
	const { ctx, cfg } = params;
	const audioConfig = cfg.tools?.media?.audio;
	if (!audioConfig || audioConfig.enabled === false) return;
	const attachments = normalizeMediaAttachments(ctx);
	if (!attachments || attachments.length === 0) return;
	const firstAudio = attachments.find((att) => att && isAudioAttachment(att) && !att.alreadyTranscribed);
	if (!firstAudio) return;
	if (shouldLogVerbose()) logVerbose(`audio-preflight: transcribing attachment ${firstAudio.index} for mention check`);
	const providerRegistry = buildProviderRegistry(params.providers);
	const cache = createMediaAttachmentCache(attachments, { localPathRoots: resolveMediaAttachmentLocalRoots({
		cfg,
		ctx
	}) });
	try {
		const result = await runCapability({
			capability: "audio",
			cfg,
			ctx,
			attachments: cache,
			media: attachments,
			agentDir: params.agentDir,
			providerRegistry,
			config: audioConfig,
			activeModel: params.activeModel
		});
		if (!result || result.outputs.length === 0) return;
		const audioOutput = result.outputs.find((output) => output.kind === "audio.transcription");
		if (!audioOutput || !audioOutput.text) return;
		firstAudio.alreadyTranscribed = true;
		if (shouldLogVerbose()) logVerbose(`audio-preflight: transcribed ${audioOutput.text.length} chars from attachment ${firstAudio.index}`);
		return audioOutput.text;
	} catch (err) {
		if (shouldLogVerbose()) logVerbose(`audio-preflight: transcription failed: ${String(err)}`);
		return;
	} finally {
		await cache.cleanup();
	}
}

//#endregion
export { transcribeFirstAudio };
