// openclaw/dist/enable-8JMVmNMv.js
import { T as normalizeChatChannelId } from "./subsystem-CYc2uK4P.js";
import { t as ensurePluginAllowlisted } from "./plugins-allowlist-D19vhOl4.js";

//#region src/plugins/enable.ts
function enablePluginInConfig(cfg, pluginId) {
	const builtInChannelId = normalizeChatChannelId(pluginId);
	const resolvedId = builtInChannelId ?? pluginId;
	if (cfg.plugins?.enabled === false) return {
		config: cfg,
		enabled: false,
		reason: "plugins disabled"
	};
	if (cfg.plugins?.deny?.includes(pluginId) || cfg.plugins?.deny?.includes(resolvedId)) return {
		config: cfg,
		enabled: false,
		reason: "blocked by denylist"
	};
	if (builtInChannelId) {
		const existing = cfg.channels?.[builtInChannelId];
		const existingRecord = existing && typeof existing === "object" && !Array.isArray(existing) ? existing : {};
		let next = {
			...cfg,
			channels: {
				...cfg.channels,
				[builtInChannelId]: {
					...existingRecord,
					enabled: true
				}
			}
		};
		next = ensurePluginAllowlisted(next, resolvedId);
		return {
			config: next,
			enabled: true
		};
	}
	const entries = {
		...cfg.plugins?.entries,
		[resolvedId]: {
			...cfg.plugins?.entries?.[resolvedId],
			enabled: true
		}
	};
	let next = {
		...cfg,
		plugins: {
			...cfg.plugins,
			entries
		}
	};
	next = ensurePluginAllowlisted(next, resolvedId);
	return {
		config: next,
		enabled: true
	};
}

//#endregion
export { enablePluginInConfig as t };
