// openclaw/dist/models-cli-CSxEBN4c.js
import { Dt as theme, v as defaultRuntime } from "./entry.js";
import "./auth-profiles-B5C565zk.js";
import "./exec-DseaG_X0.js";
import "./agent-scope-Vp6cG_dU.js";
import "./github-copilot-token-RNgXBxZS.js";
import "./host-env-security-DyQuUnEd.js";
import "./pi-auth-json-JLvgsKfX.js";
import "./model-Dckbsy27.js";
import "./pi-model-discovery-CwESh4K1.js";
import "./frontmatter-17nP3KZr.js";
import "./skills-C_gUPQn1.js";
import "./manifest-registry-CFtP_mkw.js";
import "./config-T-YRaqlE.js";
import "./env-vars-iFkEK4MO.js";
import "./dock-G4cjJyIp.js";
import "./client-Dm_agh84.js";
import "./call-BUD9fxqU.js";
import "./message-channel-CeD-0oOz.js";
import "./pairing-token-CNOCqkS5.js";
import "./net-BeVBFfmJ.js";
import "./ip-CgUPLMdA.js";
import "./tailnet-01ygwTnc.js";
import "./subagent-registry-C8AjcLWJ.js";
import "./sessions-tFJJzGVL.js";
import "./tokens-CEQAF_FI.js";
import "./plugins-mjKPZoHW.js";
import "./accounts-9Hm97-7-.js";
import "./bindings-DRU6YmSj.js";
import "./logging-CyKaqCo_.js";
import "./send-BHEev7wk.js";
import "./send--d3jcPxC.js";
import "./with-timeout-Do-sNSMI.js";
import "./deliver-jUnlfpcd.js";
import "./diagnostic-DRlvPSiu.js";
import "./diagnostic-session-state-DLuYVc4W.js";
import "./accounts-DKuYlR6W.js";
import "./send-DVX_zY9w.js";
import "./image-ops-ED4c3lgi.js";
import "./pi-embedded-helpers-B7rXYPuB.js";
import "./sandbox-DZoG3640.js";
import "./chrome-CwpCHQXf.js";
import "./tailscale-Df7MwSXG.js";
import "./auth-sHkz6y1W.js";
import "./server-context-Cjmjlel5.js";
import "./routes-FGJF5gtZ.js";
import "./redact-iUwTNBtX.js";
import "./errors-C0_m8FH_.js";
import "./fs-safe-CUG9oKdA.js";
import "./paths-B48NVbYC.js";
import "./ssrf-BUmMCCrH.js";
import "./store-DQdilc66.js";
import "./ports-Cvvcdjes.js";
import "./trash-BZfBjeIy.js";
import "./accounts-nOtv8L5i.js";
import "./paths-CXIOqFsA.js";
import "./chat-envelope-B0TExquy.js";
import "./tool-images-C_OQRH5z.js";
import "./thinking-CEjvYsJw.js";
import "./models-config-TLdOHH0m.js";
import "./exec-approvals-allowlist-DEK2b59y.js";
import "./exec-safe-bin-runtime-policy-CvZXvVH0.js";
import "./reply-prefix-Cjty59Ak.js";
import "./memory-cli-Dl01gtSM.js";
import "./manager-C7k8M9TI.js";
import "./gemini-auth-D3SYOg-4.js";
import "./fetch-guard-C8RfPptD.js";
import "./query-expansion-CpEolL3Q.js";
import "./retry-CYfnOIlD.js";
import "./target-errors-w0QNSkru.js";
import "./chunk-_4X2PcMV.js";
import "./markdown-tables-De1DeW9v.js";
import "./local-roots-rynfJ9lb.js";
import "./ir-S27Fz0Oh.js";
import "./render-kJVmJTwh.js";
import "./commands-CV3rSV1B.js";
import "./commands-registry-qeJ2eG3c.js";
import "./image-CRv24xgn.js";
import "./tool-display-DdinKh-X.js";
import "./runner-C6YJTH6y.js";
import "./model-catalog-Dw64TluO.js";
import "./session-utils-G_glLiDN.js";
import "./skill-commands-xvplpuxk.js";
import "./workspace-dirs-DOBMfIc3.js";
import "./pairing-store-CynYTGfO.js";
import "./fetch-_VotajCU.js";
import "./exec-approvals-Cur0r_mD.js";
import "./nodes-screen-BK5cXYr2.js";
import "./session-cost-usage-CsX3XzNA.js";
import "./channel-activity-DdH9i1uh.js";
import "./tables-DSqC3itu.js";
import "./control-service-qfQQoBYf.js";
import "./stagger-BJGKxryR.js";
import "./channel-selection-SYzHCfvA.js";
import "./send-CDVxGsEC.js";
import "./outbound-attachment-BNNhQo0S.js";
import "./delivery-queue-C4Eb0w7a.js";
import "./send-CMmbwphw.js";
import "./resolve-route-BnjZBagQ.js";
import "./proxy-GwQeAgiZ.js";
import { t as formatDocsLink } from "./links-CeoK7yOD.js";
import { n as runCommandWithRuntime, t as resolveOptionFromCommand } from "./cli-utils-DrfEy7kN.js";
import "./help-format-D46DMScn.js";
import "./progress-GGN1XdxY.js";
import "./replies-UOjPvhgz.js";
import "./onboard-helpers-DbKs-NAe.js";
import "./prompt-style-A8DAw4-S.js";
import "./pairing-labels-WalmCKjT.js";
import "./pi-tools.policy-D74j9leo.js";
import "./note-Di2CVd8s.js";
import "./clack-prompter-WNNLlfYu.js";
import "./table-muj3Scnd.js";
import "./shared-FeUGvfHN.js";
import "./auth-token-BjXE62L_.js";
import "./logging-COSqTUXs.js";
import { c as githubCopilotLoginCommand } from "./provider-auth-helpers-BgsX_d0R.js";
import { C as modelsAliasesListCommand, S as modelsAliasesAddCommand, _ as modelsAuthOrderSetCommand, a as modelsStatusCommand, b as modelsAuthPasteTokenCommand, c as modelsImageFallbacksClearCommand, d as modelsFallbacksAddCommand, f as modelsFallbacksClearCommand, g as modelsAuthOrderGetCommand, h as modelsAuthOrderClearCommand, i as modelsScanCommand, l as modelsImageFallbacksListCommand, m as modelsFallbacksRemoveCommand, n as modelsSetImageCommand, o as modelsListCommand, p as modelsFallbacksListCommand, r as modelsSetCommand, s as modelsImageFallbacksAddCommand, u as modelsImageFallbacksRemoveCommand, v as modelsAuthAddCommand, w as modelsAliasesRemoveCommand, x as modelsAuthSetupTokenCommand, y as modelsAuthLoginCommand } from "./models-GHAfWP6T.js";

//#region src/cli/models-cli.ts
function runModelsCommand(action) {
	return runCommandWithRuntime(defaultRuntime, action);
}
function registerModelsCli(program) {
	const models = program.command("models").description("Model discovery, scanning, and configuration").option("--status-json", "Output JSON (alias for `models status --json`)", false).option("--status-plain", "Plain output (alias for `models status --plain`)", false).option("--agent <id>", "Agent id to inspect (overrides OPENCLAW_AGENT_DIR/PI_CODING_AGENT_DIR)").addHelpText("after", () => `\n${theme.muted("Docs:")} ${formatDocsLink("/cli/models", "docs.openclaw.ai/cli/models")}\n`);
	models.command("list").description("List models (configured by default)").option("--all", "Show full model catalog", false).option("--local", "Filter to local models", false).option("--provider <name>", "Filter by provider").option("--json", "Output JSON", false).option("--plain", "Plain line output", false).action(async (opts) => {
		await runModelsCommand(async () => {
			await modelsListCommand(opts, defaultRuntime);
		});
	});
	models.command("status").description("Show configured model state").option("--json", "Output JSON", false).option("--plain", "Plain output", false).option("--check", "Exit non-zero if auth is expiring/expired (1=expired/missing, 2=expiring)", false).option("--probe", "Probe configured provider auth (live)", false).option("--probe-provider <name>", "Only probe a single provider").option("--probe-profile <id>", "Only probe specific auth profile ids (repeat or comma-separated)", (value, previous) => {
		const next = Array.isArray(previous) ? previous : previous ? [previous] : [];
		next.push(value);
		return next;
	}).option("--probe-timeout <ms>", "Per-probe timeout in ms").option("--probe-concurrency <n>", "Concurrent probes").option("--probe-max-tokens <n>", "Probe max tokens (best-effort)").option("--agent <id>", "Agent id to inspect (overrides OPENCLAW_AGENT_DIR/PI_CODING_AGENT_DIR)").action(async (opts, command) => {
		const agent = resolveOptionFromCommand(command, "agent") ?? opts.agent;
		await runModelsCommand(async () => {
			await modelsStatusCommand({
				json: Boolean(opts.json),
				plain: Boolean(opts.plain),
				check: Boolean(opts.check),
				probe: Boolean(opts.probe),
				probeProvider: opts.probeProvider,
				probeProfile: opts.probeProfile,
				probeTimeout: opts.probeTimeout,
				probeConcurrency: opts.probeConcurrency,
				probeMaxTokens: opts.probeMaxTokens,
				agent
			}, defaultRuntime);
		});
	});
	models.command("set").description("Set the default model").argument("<model>", "Model id or alias").action(async (model) => {
		await runModelsCommand(async () => {
			await modelsSetCommand(model, defaultRuntime);
		});
	});
	models.command("set-image").description("Set the image model").argument("<model>", "Model id or alias").action(async (model) => {
		await runModelsCommand(async () => {
			await modelsSetImageCommand(model, defaultRuntime);
		});
	});
	const aliases = models.command("aliases").description("Manage model aliases");
	aliases.command("list").description("List model aliases").option("--json", "Output JSON", false).option("--plain", "Plain output", false).action(async (opts) => {
		await runModelsCommand(async () => {
			await modelsAliasesListCommand(opts, defaultRuntime);
		});
	});
	aliases.command("add").description("Add or update a model alias").argument("<alias>", "Alias name").argument("<model>", "Model id or alias").action(async (alias, model) => {
		await runModelsCommand(async () => {
			await modelsAliasesAddCommand(alias, model, defaultRuntime);
		});
	});
	aliases.command("remove").description("Remove a model alias").argument("<alias>", "Alias name").action(async (alias) => {
		await runModelsCommand(async () => {
			await modelsAliasesRemoveCommand(alias, defaultRuntime);
		});
	});
	const fallbacks = models.command("fallbacks").description("Manage model fallback list");
	fallbacks.command("list").description("List fallback models").option("--json", "Output JSON", false).option("--plain", "Plain output", false).action(async (opts) => {
		await runModelsCommand(async () => {
			await modelsFallbacksListCommand(opts, defaultRuntime);
		});
	});
	fallbacks.command("add").description("Add a fallback model").argument("<model>", "Model id or alias").action(async (model) => {
		await runModelsCommand(async () => {
			await modelsFallbacksAddCommand(model, defaultRuntime);
		});
	});
	fallbacks.command("remove").description("Remove a fallback model").argument("<model>", "Model id or alias").action(async (model) => {
		await runModelsCommand(async () => {
			await modelsFallbacksRemoveCommand(model, defaultRuntime);
		});
	});
	fallbacks.command("clear").description("Clear all fallback models").action(async () => {
		await runModelsCommand(async () => {
			await modelsFallbacksClearCommand(defaultRuntime);
		});
	});
	const imageFallbacks = models.command("image-fallbacks").description("Manage image model fallback list");
	imageFallbacks.command("list").description("List image fallback models").option("--json", "Output JSON", false).option("--plain", "Plain output", false).action(async (opts) => {
		await runModelsCommand(async () => {
			await modelsImageFallbacksListCommand(opts, defaultRuntime);
		});
	});
	imageFallbacks.command("add").description("Add an image fallback model").argument("<model>", "Model id or alias").action(async (model) => {
		await runModelsCommand(async () => {
			await modelsImageFallbacksAddCommand(model, defaultRuntime);
		});
	});
	imageFallbacks.command("remove").description("Remove an image fallback model").argument("<model>", "Model id or alias").action(async (model) => {
		await runModelsCommand(async () => {
			await modelsImageFallbacksRemoveCommand(model, defaultRuntime);
		});
	});
	imageFallbacks.command("clear").description("Clear all image fallback models").action(async () => {
		await runModelsCommand(async () => {
			await modelsImageFallbacksClearCommand(defaultRuntime);
		});
	});
	models.command("scan").description("Scan OpenRouter free models for tools + images").option("--min-params <b>", "Minimum parameter size (billions)").option("--max-age-days <days>", "Skip models older than N days").option("--provider <name>", "Filter by provider prefix").option("--max-candidates <n>", "Max fallback candidates", "6").option("--timeout <ms>", "Per-probe timeout in ms").option("--concurrency <n>", "Probe concurrency").option("--no-probe", "Skip live probes; list free candidates only").option("--yes", "Accept defaults without prompting", false).option("--no-input", "Disable prompts (use defaults)").option("--set-default", "Set agents.defaults.model to the first selection", false).option("--set-image", "Set agents.defaults.imageModel to the first image selection", false).option("--json", "Output JSON", false).action(async (opts) => {
		await runModelsCommand(async () => {
			await modelsScanCommand(opts, defaultRuntime);
		});
	});
	models.action(async (opts) => {
		await runModelsCommand(async () => {
			await modelsStatusCommand({
				json: Boolean(opts?.statusJson),
				plain: Boolean(opts?.statusPlain),
				agent: opts?.agent
			}, defaultRuntime);
		});
	});
	const auth = models.command("auth").description("Manage model auth profiles");
	auth.option("--agent <id>", "Agent id for auth order get/set/clear");
	auth.action(() => {
		auth.help();
	});
	auth.command("add").description("Interactive auth helper (setup-token or paste token)").action(async () => {
		await runModelsCommand(async () => {
			await modelsAuthAddCommand({}, defaultRuntime);
		});
	});
	auth.command("login").description("Run a provider plugin auth flow (OAuth/API key)").option("--provider <id>", "Provider id registered by a plugin").option("--method <id>", "Provider auth method id").option("--set-default", "Apply the provider's default model recommendation", false).action(async (opts) => {
		await runModelsCommand(async () => {
			await modelsAuthLoginCommand({
				provider: opts.provider,
				method: opts.method,
				setDefault: Boolean(opts.setDefault)
			}, defaultRuntime);
		});
	});
	auth.command("setup-token").description("Run a provider CLI to create/sync a token (TTY required)").option("--provider <name>", "Provider id (default: anthropic)").option("--yes", "Skip confirmation", false).action(async (opts) => {
		await runModelsCommand(async () => {
			await modelsAuthSetupTokenCommand({
				provider: opts.provider,
				yes: Boolean(opts.yes)
			}, defaultRuntime);
		});
	});
	auth.command("paste-token").description("Paste a token into auth-profiles.json and update config").requiredOption("--provider <name>", "Provider id (e.g. anthropic)").option("--profile-id <id>", "Auth profile id (default: <provider>:manual)").option("--expires-in <duration>", "Optional expiry duration (e.g. 365d, 12h). Stored as absolute expiresAt.").action(async (opts) => {
		await runModelsCommand(async () => {
			await modelsAuthPasteTokenCommand({
				provider: opts.provider,
				profileId: opts.profileId,
				expiresIn: opts.expiresIn
			}, defaultRuntime);
		});
	});
	auth.command("login-github-copilot").description("Login to GitHub Copilot via GitHub device flow (TTY required)").option("--profile-id <id>", "Auth profile id (default: github-copilot:github)").option("--yes", "Overwrite existing profile without prompting", false).action(async (opts) => {
		await runModelsCommand(async () => {
			await githubCopilotLoginCommand({
				profileId: opts.profileId,
				yes: Boolean(opts.yes)
			}, defaultRuntime);
		});
	});
	const order = auth.command("order").description("Manage per-agent auth profile order overrides");
	order.command("get").description("Show per-agent auth order override (from auth-profiles.json)").requiredOption("--provider <name>", "Provider id (e.g. anthropic)").option("--agent <id>", "Agent id (default: configured default agent)").option("--json", "Output JSON", false).action(async (opts, command) => {
		const agent = resolveOptionFromCommand(command, "agent") ?? opts.agent;
		await runModelsCommand(async () => {
			await modelsAuthOrderGetCommand({
				provider: opts.provider,
				agent,
				json: Boolean(opts.json)
			}, defaultRuntime);
		});
	});
	order.command("set").description("Set per-agent auth order override (locks rotation to this list)").requiredOption("--provider <name>", "Provider id (e.g. anthropic)").option("--agent <id>", "Agent id (default: configured default agent)").argument("<profileIds...>", "Auth profile ids (e.g. anthropic:default)").action(async (profileIds, opts, command) => {
		const agent = resolveOptionFromCommand(command, "agent") ?? opts.agent;
		await runModelsCommand(async () => {
			await modelsAuthOrderSetCommand({
				provider: opts.provider,
				agent,
				order: profileIds
			}, defaultRuntime);
		});
	});
	order.command("clear").description("Clear per-agent auth order override (fall back to config/round-robin)").requiredOption("--provider <name>", "Provider id (e.g. anthropic)").option("--agent <id>", "Agent id (default: configured default agent)").action(async (opts, command) => {
		const agent = resolveOptionFromCommand(command, "agent") ?? opts.agent;
		await runModelsCommand(async () => {
			await modelsAuthOrderClearCommand({
				provider: opts.provider,
				agent
			}, defaultRuntime);
		});
	});
}

//#endregion
export { registerModelsCli };
