// openclaw/dist/register.setup-BgwImNRh.js
import "./paths-B4BZAPZh.js";
import { B as theme, S as shortenHomePath } from "./utils-DCtXtPui.js";
import { f as defaultRuntime } from "./subsystem-CYc2uK4P.js";
import "./exec-QOib7mc9.js";
import { C as ensureAgentWorkspace, m as DEFAULT_AGENT_WORKSPACE_DIR } from "./agent-scope-CJ4YDowp.js";
import "./model-selection-jLPRV2hp.js";
import "./github-copilot-token-nncItI8D.js";
import "./boolean-BgXe2hyu.js";
import "./env-Bi5nSj2_.js";
import "./host-env-security-ljCLeQmh.js";
import "./message-channel-CboOXQ5K.js";
import { l as writeConfigFile, r as createConfigIO } from "./config-D9td3nQ8.js";
import "./env-vars-CvvqezS9.js";
import "./manifest-registry-6Smo6hJJ.js";
import "./dock-CrncBkrf.js";
import "./ip-DdJWkGLX.js";
import "./tailnet-hl3MDqWZ.js";
import "./ws-CxontMkm.js";
import "./redact-DUz7OAYQ.js";
import "./errors-botak1wn.js";
import "./sessions-CX6bjL0O.js";
import "./plugins-D_vAsEMg.js";
import "./accounts-BIx69xGF.js";
import "./accounts-C-pe2tSL.js";
import "./accounts-ZHIyNWMM.js";
import "./bindings-BycbLu-x.js";
import "./logging-CeI8pPxx.js";
import { s as resolveSessionTranscriptsDir } from "./paths-CIwUV7XC.js";
import "./chat-envelope-0XVRE01I.js";
import "./client-CfOpOyvw.js";
import "./call-BV1LTlF4.js";
import "./pairing-token-DNk38FMB.js";
import { t as formatDocsLink } from "./links-Xz2s6XZA.js";
import { n as runCommandWithRuntime } from "./cli-utils-tNdOjVQI.js";
import "./progress-ebTx2ghO.js";
import "./onboard-helpers-D_b_dmpO.js";
import "./prompt-style-BBhZVg0K.js";
import "./runtime-guard-CdzxWI2g.js";
import { t as hasExplicitOptions } from "./command-options-CqFPOBqQ.js";
import "./note-D0hhOlsB.js";
import "./clack-prompter-BW4jch76.js";
import "./onboarding-CY2RYrF5.js";
import { n as logConfigUpdated, t as formatConfigPath } from "./logging-B4s0Kotf.js";
import { t as onboardCommand } from "./onboard-sj4Rk1tl.js";
import JSON5 from "json5";
import fs from "node:fs/promises";

//#region src/commands/setup.ts
async function readConfigFileRaw(configPath) {
	try {
		const raw = await fs.readFile(configPath, "utf-8");
		const parsed = JSON5.parse(raw);
		if (parsed && typeof parsed === "object") return {
			exists: true,
			parsed
		};
		return {
			exists: true,
			parsed: {}
		};
	} catch {
		return {
			exists: false,
			parsed: {}
		};
	}
}
async function setupCommand(opts, runtime = defaultRuntime) {
	const desiredWorkspace = typeof opts?.workspace === "string" && opts.workspace.trim() ? opts.workspace.trim() : void 0;
	const configPath = createConfigIO().configPath;
	const existingRaw = await readConfigFileRaw(configPath);
	const cfg = existingRaw.parsed;
	const defaults = cfg.agents?.defaults ?? {};
	const workspace = desiredWorkspace ?? defaults.workspace ?? DEFAULT_AGENT_WORKSPACE_DIR;
	const next = {
		...cfg,
		agents: {
			...cfg.agents,
			defaults: {
				...defaults,
				workspace
			}
		}
	};
	if (!existingRaw.exists || defaults.workspace !== workspace) {
		await writeConfigFile(next);
		if (!existingRaw.exists) runtime.log(`Wrote ${formatConfigPath(configPath)}`);
		else logConfigUpdated(runtime, {
			path: configPath,
			suffix: "(set agents.defaults.workspace)"
		});
	} else runtime.log(`Config OK: ${formatConfigPath(configPath)}`);
	const ws = await ensureAgentWorkspace({
		dir: workspace,
		ensureBootstrapFiles: !next.agents?.defaults?.skipBootstrap
	});
	runtime.log(`Workspace OK: ${shortenHomePath(ws.dir)}`);
	const sessionsDir = resolveSessionTranscriptsDir();
	await fs.mkdir(sessionsDir, { recursive: true });
	runtime.log(`Sessions OK: ${shortenHomePath(sessionsDir)}`);
}

//#endregion
//#region src/cli/program/register.setup.ts
function registerSetupCommand(program) {
	program.command("setup").description("Initialize ~/.openclaw/openclaw.json and the agent workspace").addHelpText("after", () => `\n${theme.muted("Docs:")} ${formatDocsLink("/cli/setup", "docs.openclaw.ai/cli/setup")}\n`).option("--workspace <dir>", "Agent workspace directory (default: ~/.openclaw/workspace; stored as agents.defaults.workspace)").option("--wizard", "Run the interactive onboarding wizard", false).option("--non-interactive", "Run the wizard without prompts", false).option("--mode <mode>", "Wizard mode: local|remote").option("--remote-url <url>", "Remote Gateway WebSocket URL").option("--remote-token <token>", "Remote Gateway token (optional)").action(async (opts, command) => {
		await runCommandWithRuntime(defaultRuntime, async () => {
			const hasWizardFlags = hasExplicitOptions(command, [
				"wizard",
				"nonInteractive",
				"mode",
				"remoteUrl",
				"remoteToken"
			]);
			if (opts.wizard || hasWizardFlags) {
				await onboardCommand({
					workspace: opts.workspace,
					nonInteractive: Boolean(opts.nonInteractive),
					mode: opts.mode,
					remoteUrl: opts.remoteUrl,
					remoteToken: opts.remoteToken
				}, defaultRuntime);
				return;
			}
			await setupCommand({ workspace: opts.workspace }, defaultRuntime);
		});
	});
}

//#endregion
export { registerSetupCommand };
