// openclaw/dist/register.status-health-sessions-DPCGgBxM.js
import { Dt as theme, v as defaultRuntime, xt as setVerbose } from "./entry.js";
import "./auth-profiles-B5C565zk.js";
import "./exec-DseaG_X0.js";
import "./agent-scope-Vp6cG_dU.js";
import "./github-copilot-token-RNgXBxZS.js";
import "./host-env-security-DyQuUnEd.js";
import "./model-Dckbsy27.js";
import "./pi-model-discovery-CwESh4K1.js";
import "./frontmatter-17nP3KZr.js";
import "./skills-C_gUPQn1.js";
import "./manifest-registry-CFtP_mkw.js";
import "./skills-status-BNEhWdVQ.js";
import "./config-T-YRaqlE.js";
import "./env-vars-iFkEK4MO.js";
import "./dock-G4cjJyIp.js";
import "./client-Dm_agh84.js";
import "./call-BUD9fxqU.js";
import "./message-channel-CeD-0oOz.js";
import "./pairing-token-CNOCqkS5.js";
import "./net-BeVBFfmJ.js";
import "./ip-CgUPLMdA.js";
import "./tailnet-01ygwTnc.js";
import "./subagent-registry-C8AjcLWJ.js";
import "./sessions-tFJJzGVL.js";
import "./tokens-CEQAF_FI.js";
import "./plugins-mjKPZoHW.js";
import "./accounts-9Hm97-7-.js";
import "./bindings-DRU6YmSj.js";
import "./logging-CyKaqCo_.js";
import "./send-BHEev7wk.js";
import "./send--d3jcPxC.js";
import "./with-timeout-Do-sNSMI.js";
import "./deliver-jUnlfpcd.js";
import "./diagnostic-DRlvPSiu.js";
import "./diagnostic-session-state-DLuYVc4W.js";
import "./accounts-DKuYlR6W.js";
import "./send-DVX_zY9w.js";
import "./image-ops-ED4c3lgi.js";
import "./pi-embedded-helpers-B7rXYPuB.js";
import "./sandbox-DZoG3640.js";
import "./chrome-CwpCHQXf.js";
import "./tailscale-Df7MwSXG.js";
import "./auth-sHkz6y1W.js";
import "./server-context-Cjmjlel5.js";
import "./routes-FGJF5gtZ.js";
import "./redact-iUwTNBtX.js";
import "./errors-C0_m8FH_.js";
import "./fs-safe-CUG9oKdA.js";
import "./paths-B48NVbYC.js";
import "./ssrf-BUmMCCrH.js";
import "./store-DQdilc66.js";
import "./ports-Cvvcdjes.js";
import "./trash-BZfBjeIy.js";
import "./accounts-nOtv8L5i.js";
import "./paths-CXIOqFsA.js";
import "./chat-envelope-B0TExquy.js";
import "./tool-images-C_OQRH5z.js";
import "./thinking-CEjvYsJw.js";
import "./models-config-TLdOHH0m.js";
import "./exec-approvals-allowlist-DEK2b59y.js";
import "./exec-safe-bin-runtime-policy-CvZXvVH0.js";
import "./reply-prefix-Cjty59Ak.js";
import "./memory-cli-Dl01gtSM.js";
import "./manager-C7k8M9TI.js";
import "./gemini-auth-D3SYOg-4.js";
import "./fetch-guard-C8RfPptD.js";
import "./query-expansion-CpEolL3Q.js";
import "./retry-CYfnOIlD.js";
import "./target-errors-w0QNSkru.js";
import "./chunk-_4X2PcMV.js";
import "./markdown-tables-De1DeW9v.js";
import "./local-roots-rynfJ9lb.js";
import "./ir-S27Fz0Oh.js";
import "./render-kJVmJTwh.js";
import "./commands-CV3rSV1B.js";
import "./commands-registry-qeJ2eG3c.js";
import "./image-CRv24xgn.js";
import "./tool-display-DdinKh-X.js";
import "./runner-C6YJTH6y.js";
import "./model-catalog-Dw64TluO.js";
import "./session-utils-G_glLiDN.js";
import "./skill-commands-xvplpuxk.js";
import "./workspace-dirs-DOBMfIc3.js";
import "./pairing-store-CynYTGfO.js";
import "./fetch-_VotajCU.js";
import "./exec-approvals-Cur0r_mD.js";
import "./nodes-screen-BK5cXYr2.js";
import "./session-cost-usage-CsX3XzNA.js";
import "./channel-activity-DdH9i1uh.js";
import "./tables-DSqC3itu.js";
import "./control-service-qfQQoBYf.js";
import "./stagger-BJGKxryR.js";
import "./channel-selection-SYzHCfvA.js";
import "./send-CDVxGsEC.js";
import "./outbound-attachment-BNNhQo0S.js";
import "./delivery-queue-C4Eb0w7a.js";
import "./send-CMmbwphw.js";
import "./resolve-route-BnjZBagQ.js";
import "./proxy-GwQeAgiZ.js";
import { t as formatDocsLink } from "./links-CeoK7yOD.js";
import { n as runCommandWithRuntime } from "./cli-utils-DrfEy7kN.js";
import { t as formatHelpExamples } from "./help-format-D46DMScn.js";
import "./progress-GGN1XdxY.js";
import "./replies-UOjPvhgz.js";
import "./onboard-helpers-DbKs-NAe.js";
import "./prompt-style-A8DAw4-S.js";
import "./pairing-labels-WalmCKjT.js";
import "./pi-tools.policy-D74j9leo.js";
import "./dangerous-tools-CdTnu3m4.js";
import "./skill-scanner-BexBfR1i.js";
import "./channels-status-issues-3x6ZUz93.js";
import { n as parsePositiveIntOrUndefined } from "./helpers-D3MFXmv_.js";
import "./runtime-guard-DXZzkWDN.js";
import "./systemd-C8J0px_u.js";
import "./service-BLDuAGNF.js";
import "./diagnostics-gBXKmaH8.js";
import "./table-muj3Scnd.js";
import "./audit-CiNk972u.js";
import { n as statusCommand } from "./status-DU8tDKsE.js";
import { r as healthCommand } from "./health-tb27M5Ss.js";
import "./channel-account-context-CZ4WBsul.js";
import "./dm-policy-shared-pGa8zQCc.js";
import "./node-service-D-I_FGIb.js";
import "./status.update-ZCkF20eu.js";
import { t as sessionsCommand } from "./sessions-C4M_C4hb.js";

//#region src/cli/program/register.status-health-sessions.ts
function resolveVerbose(opts) {
	return Boolean(opts.verbose || opts.debug);
}
function parseTimeoutMs(timeout) {
	const parsed = parsePositiveIntOrUndefined(timeout);
	if (timeout !== void 0 && parsed === void 0) {
		defaultRuntime.error("--timeout must be a positive integer (milliseconds)");
		defaultRuntime.exit(1);
		return null;
	}
	return parsed;
}
async function runWithVerboseAndTimeout(opts, action) {
	const verbose = resolveVerbose(opts);
	setVerbose(verbose);
	const timeoutMs = parseTimeoutMs(opts.timeout);
	if (timeoutMs === null) return;
	await runCommandWithRuntime(defaultRuntime, async () => {
		await action({
			verbose,
			timeoutMs
		});
	});
}
function registerStatusHealthSessionsCommands(program) {
	program.command("status").description("Show channel health and recent session recipients").option("--json", "Output JSON instead of text", false).option("--all", "Full diagnosis (read-only, pasteable)", false).option("--usage", "Show model provider usage/quota snapshots", false).option("--deep", "Probe channels (WhatsApp Web + Telegram + Discord + Slack + Signal)", false).option("--timeout <ms>", "Probe timeout in milliseconds", "10000").option("--verbose", "Verbose logging", false).option("--debug", "Alias for --verbose", false).addHelpText("after", () => `\n${theme.heading("Examples:")}\n${formatHelpExamples([
		["openclaw status", "Show channel health + session summary."],
		["openclaw status --all", "Full diagnosis (read-only)."],
		["openclaw status --json", "Machine-readable output."],
		["openclaw status --usage", "Show model provider usage/quota snapshots."],
		["openclaw status --deep", "Run channel probes (WA + Telegram + Discord + Slack + Signal)."],
		["openclaw status --deep --timeout 5000", "Tighten probe timeout."]
	])}`).addHelpText("after", () => `\n${theme.muted("Docs:")} ${formatDocsLink("/cli/status", "docs.openclaw.ai/cli/status")}\n`).action(async (opts) => {
		await runWithVerboseAndTimeout(opts, async ({ verbose, timeoutMs }) => {
			await statusCommand({
				json: Boolean(opts.json),
				all: Boolean(opts.all),
				deep: Boolean(opts.deep),
				usage: Boolean(opts.usage),
				timeoutMs,
				verbose
			}, defaultRuntime);
		});
	});
	program.command("health").description("Fetch health from the running gateway").option("--json", "Output JSON instead of text", false).option("--timeout <ms>", "Connection timeout in milliseconds", "10000").option("--verbose", "Verbose logging", false).option("--debug", "Alias for --verbose", false).addHelpText("after", () => `\n${theme.muted("Docs:")} ${formatDocsLink("/cli/health", "docs.openclaw.ai/cli/health")}\n`).action(async (opts) => {
		await runWithVerboseAndTimeout(opts, async ({ verbose, timeoutMs }) => {
			await healthCommand({
				json: Boolean(opts.json),
				timeoutMs,
				verbose
			}, defaultRuntime);
		});
	});
	program.command("sessions").description("List stored conversation sessions").option("--json", "Output as JSON", false).option("--verbose", "Verbose logging", false).option("--store <path>", "Path to session store (default: resolved from config)").option("--active <minutes>", "Only show sessions updated within the past N minutes").addHelpText("after", () => `\n${theme.heading("Examples:")}\n${formatHelpExamples([
		["openclaw sessions", "List all sessions."],
		["openclaw sessions --active 120", "Only last 2 hours."],
		["openclaw sessions --json", "Machine-readable output."],
		["openclaw sessions --store ./tmp/sessions.json", "Use a specific session store."]
	])}\n\n${theme.muted("Shows token usage per session when the agent reports it; set agents.defaults.contextTokens to cap the window and show %.")}`).addHelpText("after", () => `\n${theme.muted("Docs:")} ${formatDocsLink("/cli/sessions", "docs.openclaw.ai/cli/sessions")}\n`).action(async (opts) => {
		setVerbose(Boolean(opts.verbose));
		await sessionsCommand({
			json: Boolean(opts.json),
			store: opts.store,
			active: opts.active
		}, defaultRuntime);
	});
}

//#endregion
export { registerStatusHealthSessionsCommands };
