// openclaw/dist/run-main-DzEReUj6.js
import { a as normalizeEnv, an as getCommandPath, cn as getPrimaryCommand, dn as hasHelpOrVersion, l as enableConsoleCapture, ln as getVerboseFlag, on as getFlagValue, r as isTruthyEnvValue, rn as normalizeWindowsArgv, sn as getPositiveIntFlagValue, un as hasFlag, v as defaultRuntime } from "./entry.js";
import "./auth-profiles-B5C565zk.js";
import "./exec-DseaG_X0.js";
import "./agent-scope-Vp6cG_dU.js";
import "./github-copilot-token-RNgXBxZS.js";
import "./host-env-security-DyQuUnEd.js";
import "./model-Dckbsy27.js";
import "./pi-model-discovery-CwESh4K1.js";
import "./frontmatter-17nP3KZr.js";
import "./skills-C_gUPQn1.js";
import "./manifest-registry-CFtP_mkw.js";
import { wt as loadDotEnv } from "./config-T-YRaqlE.js";
import { r as VERSION } from "./env-vars-iFkEK4MO.js";
import "./dock-G4cjJyIp.js";
import "./client-Dm_agh84.js";
import "./call-BUD9fxqU.js";
import "./message-channel-CeD-0oOz.js";
import "./pairing-token-CNOCqkS5.js";
import "./net-BeVBFfmJ.js";
import "./ip-CgUPLMdA.js";
import "./tailnet-01ygwTnc.js";
import "./subagent-registry-C8AjcLWJ.js";
import "./sessions-tFJJzGVL.js";
import "./tokens-CEQAF_FI.js";
import "./plugins-mjKPZoHW.js";
import "./accounts-9Hm97-7-.js";
import "./bindings-DRU6YmSj.js";
import "./logging-CyKaqCo_.js";
import "./send-BHEev7wk.js";
import "./send--d3jcPxC.js";
import "./with-timeout-Do-sNSMI.js";
import "./deliver-jUnlfpcd.js";
import "./diagnostic-DRlvPSiu.js";
import "./diagnostic-session-state-DLuYVc4W.js";
import "./accounts-DKuYlR6W.js";
import "./send-DVX_zY9w.js";
import "./image-ops-ED4c3lgi.js";
import "./pi-embedded-helpers-B7rXYPuB.js";
import "./sandbox-DZoG3640.js";
import "./chrome-CwpCHQXf.js";
import "./tailscale-Df7MwSXG.js";
import "./auth-sHkz6y1W.js";
import "./server-context-Cjmjlel5.js";
import "./routes-FGJF5gtZ.js";
import "./redact-iUwTNBtX.js";
import { r as formatUncaughtError } from "./errors-C0_m8FH_.js";
import "./fs-safe-CUG9oKdA.js";
import "./paths-B48NVbYC.js";
import "./ssrf-BUmMCCrH.js";
import "./store-DQdilc66.js";
import "./ports-Cvvcdjes.js";
import "./trash-BZfBjeIy.js";
import "./accounts-nOtv8L5i.js";
import "./paths-CXIOqFsA.js";
import "./chat-envelope-B0TExquy.js";
import "./tool-images-C_OQRH5z.js";
import "./thinking-CEjvYsJw.js";
import "./models-config-TLdOHH0m.js";
import "./exec-approvals-allowlist-DEK2b59y.js";
import "./exec-safe-bin-runtime-policy-CvZXvVH0.js";
import "./reply-prefix-Cjty59Ak.js";
import "./memory-cli-Dl01gtSM.js";
import "./manager-C7k8M9TI.js";
import "./gemini-auth-D3SYOg-4.js";
import "./fetch-guard-C8RfPptD.js";
import "./query-expansion-CpEolL3Q.js";
import "./retry-CYfnOIlD.js";
import "./target-errors-w0QNSkru.js";
import "./chunk-_4X2PcMV.js";
import "./markdown-tables-De1DeW9v.js";
import "./local-roots-rynfJ9lb.js";
import "./ir-S27Fz0Oh.js";
import "./render-kJVmJTwh.js";
import "./commands-CV3rSV1B.js";
import "./commands-registry-qeJ2eG3c.js";
import "./image-CRv24xgn.js";
import "./tool-display-DdinKh-X.js";
import { u as installUnhandledRejectionHandler } from "./runner-C6YJTH6y.js";
import "./model-catalog-Dw64TluO.js";
import "./session-utils-G_glLiDN.js";
import "./skill-commands-xvplpuxk.js";
import "./workspace-dirs-DOBMfIc3.js";
import "./pairing-store-CynYTGfO.js";
import "./fetch-_VotajCU.js";
import "./exec-approvals-Cur0r_mD.js";
import "./nodes-screen-BK5cXYr2.js";
import "./session-cost-usage-CsX3XzNA.js";
import "./channel-activity-DdH9i1uh.js";
import "./tables-DSqC3itu.js";
import "./control-service-qfQQoBYf.js";
import "./stagger-BJGKxryR.js";
import "./channel-selection-SYzHCfvA.js";
import "./send-CDVxGsEC.js";
import "./outbound-attachment-BNNhQo0S.js";
import "./delivery-queue-C4Eb0w7a.js";
import "./send-CMmbwphw.js";
import "./resolve-route-BnjZBagQ.js";
import "./proxy-GwQeAgiZ.js";
import "./links-CeoK7yOD.js";
import "./cli-utils-DrfEy7kN.js";
import "./help-format-D46DMScn.js";
import "./progress-GGN1XdxY.js";
import "./replies-UOjPvhgz.js";
import "./onboard-helpers-DbKs-NAe.js";
import "./prompt-style-A8DAw4-S.js";
import "./pairing-labels-WalmCKjT.js";
import "./pi-tools.policy-D74j9leo.js";
import { t as ensureOpenClawCliOnPath } from "./path-env-_hyEyMcT.js";
import "./catalog-Csdt88Yh.js";
import "./note-Di2CVd8s.js";
import "./plugin-auto-enable-DqtMi0yW.js";
import { t as ensurePluginRegistryLoaded } from "./plugin-registry-CNdNJmcU.js";
import { t as assertSupportedRuntime } from "./runtime-guard-DXZzkWDN.js";
import { t as emitCliBanner } from "./banner-FT3hiPuT.js";
import "./doctor-config-flow-Chhpi_J6.js";
import { n as ensureConfigReady } from "./config-guard-BJut8dwf.js";
import process$1 from "node:process";
import { fileURLToPath } from "node:url";

//#region src/cli/program/routes.ts
const routeHealth = {
	match: (path) => path[0] === "health",
	loadPlugins: true,
	run: async (argv) => {
		const json = hasFlag(argv, "--json");
		const verbose = getVerboseFlag(argv, { includeDebug: true });
		const timeoutMs = getPositiveIntFlagValue(argv, "--timeout");
		if (timeoutMs === null) return false;
		const { healthCommand } = await import("./health-tb27M5Ss.js").then((n) => n.i);
		await healthCommand({
			json,
			timeoutMs,
			verbose
		}, defaultRuntime);
		return true;
	}
};
const routeStatus = {
	match: (path) => path[0] === "status",
	loadPlugins: true,
	run: async (argv) => {
		const json = hasFlag(argv, "--json");
		const deep = hasFlag(argv, "--deep");
		const all = hasFlag(argv, "--all");
		const usage = hasFlag(argv, "--usage");
		const verbose = getVerboseFlag(argv, { includeDebug: true });
		const timeoutMs = getPositiveIntFlagValue(argv, "--timeout");
		if (timeoutMs === null) return false;
		const { statusCommand } = await import("./status-DU8tDKsE.js").then((n) => n.t);
		await statusCommand({
			json,
			deep,
			all,
			usage,
			timeoutMs,
			verbose
		}, defaultRuntime);
		return true;
	}
};
const routeSessions = {
	match: (path) => path[0] === "sessions",
	run: async (argv) => {
		const json = hasFlag(argv, "--json");
		const store = getFlagValue(argv, "--store");
		if (store === null) return false;
		const active = getFlagValue(argv, "--active");
		if (active === null) return false;
		const { sessionsCommand } = await import("./sessions-C4M_C4hb.js").then((n) => n.n);
		await sessionsCommand({
			json,
			store,
			active
		}, defaultRuntime);
		return true;
	}
};
const routeAgentsList = {
	match: (path) => path[0] === "agents" && path[1] === "list",
	run: async (argv) => {
		const json = hasFlag(argv, "--json");
		const bindings = hasFlag(argv, "--bindings");
		const { agentsListCommand } = await import("./agents-IGgEPv3P.js").then((n) => n.t);
		await agentsListCommand({
			json,
			bindings
		}, defaultRuntime);
		return true;
	}
};
const routeMemoryStatus = {
	match: (path) => path[0] === "memory" && path[1] === "status",
	run: async (argv) => {
		const agent = getFlagValue(argv, "--agent");
		if (agent === null) return false;
		const json = hasFlag(argv, "--json");
		const deep = hasFlag(argv, "--deep");
		const index = hasFlag(argv, "--index");
		const verbose = hasFlag(argv, "--verbose");
		const { runMemoryStatus } = await import("./memory-cli-Dl01gtSM.js").then((n) => n.t);
		await runMemoryStatus({
			agent,
			json,
			deep,
			index,
			verbose
		});
		return true;
	}
};
function getCommandPositionals(argv) {
	const out = [];
	const args = argv.slice(2);
	for (const arg of args) {
		if (!arg || arg === "--") break;
		if (arg.startsWith("-")) continue;
		out.push(arg);
	}
	return out;
}
function getFlagValues(argv, name) {
	const values = [];
	const args = argv.slice(2);
	for (let i = 0; i < args.length; i += 1) {
		const arg = args[i];
		if (!arg || arg === "--") break;
		if (arg === name) {
			const next = args[i + 1];
			if (!next || next === "--" || next.startsWith("-")) return null;
			values.push(next);
			i += 1;
			continue;
		}
		if (arg.startsWith(`${name}=`)) {
			const value = arg.slice(name.length + 1).trim();
			if (!value) return null;
			values.push(value);
		}
	}
	return values;
}
const routes = [
	routeHealth,
	routeStatus,
	routeSessions,
	routeAgentsList,
	routeMemoryStatus,
	{
		match: (path) => path[0] === "config" && path[1] === "get",
		run: async (argv) => {
			const pathArg = getCommandPositionals(argv)[2];
			if (!pathArg) return false;
			const json = hasFlag(argv, "--json");
			const { runConfigGet } = await import("./config-cli-B4p0hFOw.js");
			await runConfigGet({
				path: pathArg,
				json
			});
			return true;
		}
	},
	{
		match: (path) => path[0] === "config" && path[1] === "unset",
		run: async (argv) => {
			const pathArg = getCommandPositionals(argv)[2];
			if (!pathArg) return false;
			const { runConfigUnset } = await import("./config-cli-B4p0hFOw.js");
			await runConfigUnset({ path: pathArg });
			return true;
		}
	},
	{
		match: (path) => path[0] === "models" && path[1] === "list",
		run: async (argv) => {
			const provider = getFlagValue(argv, "--provider");
			if (provider === null) return false;
			const all = hasFlag(argv, "--all");
			const local = hasFlag(argv, "--local");
			const json = hasFlag(argv, "--json");
			const plain = hasFlag(argv, "--plain");
			const { modelsListCommand } = await import("./models-GHAfWP6T.js").then((n) => n.t);
			await modelsListCommand({
				all,
				local,
				provider,
				json,
				plain
			}, defaultRuntime);
			return true;
		}
	},
	{
		match: (path) => path[0] === "models" && path[1] === "status",
		run: async (argv) => {
			const probeProvider = getFlagValue(argv, "--probe-provider");
			if (probeProvider === null) return false;
			const probeTimeout = getFlagValue(argv, "--probe-timeout");
			if (probeTimeout === null) return false;
			const probeConcurrency = getFlagValue(argv, "--probe-concurrency");
			if (probeConcurrency === null) return false;
			const probeMaxTokens = getFlagValue(argv, "--probe-max-tokens");
			if (probeMaxTokens === null) return false;
			const agent = getFlagValue(argv, "--agent");
			if (agent === null) return false;
			const probeProfileValues = getFlagValues(argv, "--probe-profile");
			if (probeProfileValues === null) return false;
			const probeProfile = probeProfileValues.length === 0 ? void 0 : probeProfileValues.length === 1 ? probeProfileValues[0] : probeProfileValues;
			const json = hasFlag(argv, "--json");
			const plain = hasFlag(argv, "--plain");
			const check = hasFlag(argv, "--check");
			const probe = hasFlag(argv, "--probe");
			const { modelsStatusCommand } = await import("./models-GHAfWP6T.js").then((n) => n.t);
			await modelsStatusCommand({
				json,
				plain,
				check,
				probe,
				probeProvider,
				probeProfile,
				probeTimeout,
				probeConcurrency,
				probeMaxTokens,
				agent
			}, defaultRuntime);
			return true;
		}
	}
];
function findRoutedCommand(path) {
	for (const route of routes) if (route.match(path)) return route;
	return null;
}

//#endregion
//#region src/cli/route.ts
async function prepareRoutedCommand(params) {
	emitCliBanner(VERSION, { argv: params.argv });
	await ensureConfigReady({
		runtime: defaultRuntime,
		commandPath: params.commandPath
	});
	if (params.loadPlugins) ensurePluginRegistryLoaded();
}
async function tryRouteCli(argv) {
	if (isTruthyEnvValue(process.env.OPENCLAW_DISABLE_ROUTE_FIRST)) return false;
	if (hasHelpOrVersion(argv)) return false;
	const path = getCommandPath(argv, 2);
	if (!path[0]) return false;
	const route = findRoutedCommand(path);
	if (!route) return false;
	await prepareRoutedCommand({
		argv,
		commandPath: path,
		loadPlugins: route.loadPlugins
	});
	return route.run(argv);
}

//#endregion
//#region src/cli/run-main.ts
function rewriteUpdateFlagArgv(argv) {
	const index = argv.indexOf("--update");
	if (index === -1) return argv;
	const next = [...argv];
	next.splice(index, 1, "update");
	return next;
}
function shouldSkipPluginCommandRegistration(params) {
	if (params.hasBuiltinPrimary) return true;
	if (!params.primary) return hasHelpOrVersion(params.argv);
	return false;
}
function shouldEnsureCliPath(argv) {
	if (hasHelpOrVersion(argv)) return false;
	const [primary, secondary] = getCommandPath(argv, 2);
	if (!primary) return true;
	if (primary === "status" || primary === "health" || primary === "sessions") return false;
	if (primary === "config" && (secondary === "get" || secondary === "unset")) return false;
	if (primary === "models" && (secondary === "list" || secondary === "status")) return false;
	return true;
}
async function runCli(argv = process$1.argv) {
	const normalizedArgv = normalizeWindowsArgv(argv);
	loadDotEnv({ quiet: true });
	normalizeEnv();
	if (shouldEnsureCliPath(normalizedArgv)) ensureOpenClawCliOnPath();
	assertSupportedRuntime();
	if (await tryRouteCli(normalizedArgv)) return;
	enableConsoleCapture();
	const { buildProgram } = await import("./program-DBVkeJo8.js");
	const program = buildProgram();
	installUnhandledRejectionHandler();
	process$1.on("uncaughtException", (error) => {
		console.error("[openclaw] Uncaught exception:", formatUncaughtError(error));
		process$1.exit(1);
	});
	const parseArgv = rewriteUpdateFlagArgv(normalizedArgv);
	const primary = getPrimaryCommand(parseArgv);
	if (primary) {
		const { getProgramContext } = await import("./program-context-Jplgkt2G.js").then((n) => n.n);
		const ctx = getProgramContext(program);
		if (ctx) {
			const { registerCoreCliByName } = await import("./command-registry-CBy4ej_0.js").then((n) => n.t);
			await registerCoreCliByName(program, ctx, primary, parseArgv);
		}
		const { registerSubCliByName } = await import("./register.subclis-Cqp8ay9s.js").then((n) => n.a);
		await registerSubCliByName(program, primary);
	}
	if (!shouldSkipPluginCommandRegistration({
		argv: parseArgv,
		primary,
		hasBuiltinPrimary: primary !== null && program.commands.some((command) => command.name() === primary)
	})) {
		const { registerPluginCliCommands } = await import("./cli-DWyvMnDJ.js");
		const { loadConfig } = await import("./config-T-YRaqlE.js").then((n) => n.t);
		registerPluginCliCommands(program, loadConfig());
	}
	await program.parseAsync(parseArgv);
}

//#endregion
export { runCli };
