// openclaw/dist/tui-cli-fyE4-nD2.js
import "./paths-B4BZAPZh.js";
import { B as theme } from "./utils-DCtXtPui.js";
import "./thinking-EAliFiVK.js";
import { f as defaultRuntime } from "./subsystem-CYc2uK4P.js";
import "./exec-QOib7mc9.js";
import "./agent-scope-CJ4YDowp.js";
import "./model-selection-jLPRV2hp.js";
import "./github-copilot-token-nncItI8D.js";
import "./boolean-BgXe2hyu.js";
import "./env-Bi5nSj2_.js";
import "./host-env-security-ljCLeQmh.js";
import "./message-channel-CboOXQ5K.js";
import "./config-D9td3nQ8.js";
import "./env-vars-CvvqezS9.js";
import "./manifest-registry-6Smo6hJJ.js";
import "./dock-CrncBkrf.js";
import "./pi-embedded-helpers-husLPufc.js";
import "./sandbox-BfRAzU9M.js";
import "./chrome-DFWJJMOL.js";
import "./tailscale-BQzlD90q.js";
import "./ip-DdJWkGLX.js";
import "./tailnet-hl3MDqWZ.js";
import "./ws-CxontMkm.js";
import "./auth-lSe2vcAf.js";
import "./server-context-BLGh1cTt.js";
import "./frontmatter-CDXvVVtB.js";
import "./skills-DikywEzc.js";
import "./routes-CmNAokG-.js";
import "./redact-DUz7OAYQ.js";
import "./errors-botak1wn.js";
import "./fs-safe-g1vUuj4t.js";
import "./paths-Bt1FN5h-.js";
import "./ssrf-CTVJWS7u.js";
import "./image-ops-CBA3OF6h.js";
import "./store-DLffydGb.js";
import "./ports-Bnye_-N3.js";
import "./trash-DjbQC3BX.js";
import "./sessions-CX6bjL0O.js";
import "./plugins-D_vAsEMg.js";
import "./accounts-BIx69xGF.js";
import "./accounts-C-pe2tSL.js";
import "./accounts-ZHIyNWMM.js";
import "./bindings-BycbLu-x.js";
import "./logging-CeI8pPxx.js";
import "./paths-CIwUV7XC.js";
import "./chat-envelope-0XVRE01I.js";
import "./tool-images-DUVex7kP.js";
import "./tool-display-C35ig9vC.js";
import "./commands-xtsUE-jY.js";
import "./commands-registry-BAhtj8hc.js";
import "./client-CfOpOyvw.js";
import "./call-BV1LTlF4.js";
import "./pairing-token-DNk38FMB.js";
import { t as formatDocsLink } from "./links-Xz2s6XZA.js";
import { t as parseTimeoutMs } from "./parse-timeout-DCQW0IQP.js";
import { t as runTui } from "./tui-DOArcj-x.js";

//#region src/cli/tui-cli.ts
function registerTuiCli(program) {
	program.command("tui").description("Open a terminal UI connected to the Gateway").option("--url <url>", "Gateway WebSocket URL (defaults to gateway.remote.url when configured)").option("--token <token>", "Gateway token (if required)").option("--password <password>", "Gateway password (if required)").option("--session <key>", "Session key (default: \"main\", or \"global\" when scope is global)").option("--deliver", "Deliver assistant replies", false).option("--thinking <level>", "Thinking level override").option("--message <text>", "Send an initial message after connecting").option("--timeout-ms <ms>", "Agent timeout in ms (defaults to agents.defaults.timeoutSeconds)").option("--history-limit <n>", "History entries to load", "200").addHelpText("after", () => `\n${theme.muted("Docs:")} ${formatDocsLink("/cli/tui", "docs.openclaw.ai/cli/tui")}\n`).action(async (opts) => {
		try {
			const timeoutMs = parseTimeoutMs(opts.timeoutMs);
			if (opts.timeoutMs !== void 0 && timeoutMs === void 0) defaultRuntime.error(`warning: invalid --timeout-ms "${String(opts.timeoutMs)}"; ignoring`);
			const historyLimit = Number.parseInt(String(opts.historyLimit ?? "200"), 10);
			await runTui({
				url: opts.url,
				token: opts.token,
				password: opts.password,
				session: opts.session,
				deliver: Boolean(opts.deliver),
				thinking: opts.thinking,
				message: opts.message,
				timeoutMs,
				historyLimit: Number.isNaN(historyLimit) ? void 0 : historyLimit
			});
		} catch (err) {
			defaultRuntime.error(String(err));
			defaultRuntime.exit(1);
		}
	});
}

//#endregion
export { registerTuiCli };
