// openclaw/dist/plugin-sdk/audio-preflight-fgh84Ci0.js
import "./accounts-D-nCmJLC.js";
import "./paths-DVWx7USN.js";
import "./github-copilot-token-Cg0YPPSu.js";
import "./plugins-B5m5N2-_.js";
import { Z as logVerbose, et as shouldLogVerbose } from "./subsystem-B3gKmYuz.js";
import "./config-D7yhe8WC.js";
import "./command-format-Ck33m20r.js";
import "./model-selection-Be0rkSSc.js";
import "./agent-scope-DeS5VuCp.js";
import "./manifest-registry-DNqjKxg2.js";
import "./dock-B-ZHp6hI.js";
import "./redact-4KGGYfvj.js";
import "./errors-1rgxcvGb.js";
import "./image-ops-CZuX7Z1C.js";
import "./ssrf-DswcqIq6.js";
import "./fetch-guard-DDo5yfUq.js";
import "./local-roots-g06aP0A4.js";
import "./message-channel-1SjHkh9S.js";
import "./bindings-DFv2Dwnp.js";
import "./tool-images-D8grjE7E.js";
import { a as resolveMediaAttachmentLocalRoots, n as createMediaAttachmentCache, o as runCapability, r as normalizeMediaAttachments, t as buildProviderRegistry, u as isAudioAttachment } from "./runner-CrEoy4qC.js";
import "./skills-Gt9vbHJR.js";
import "./chrome-B6OIe2xL.js";
import "./accounts-B6MN4-iI.js";
import "./accounts-DInAeEqh.js";
import "./sessions-Kufe0AiW.js";
import "./paths-KtBBYvS4.js";
import "./store-CXYl4NNL.js";
import "./pi-embedded-helpers-B_jcTbpF.js";
import "./thinking-DFs8hJTB.js";
import "./image-DvMwSfaa.js";
import "./pi-model-discovery-CLdAHc9I.js";
import "./api-key-rotation-DtqAmYo3.js";

//#region src/media-understanding/audio-preflight.ts
/**
* Transcribes the first audio attachment BEFORE mention checking.
* This allows voice notes to be processed in group chats with requireMention: true.
* Returns the transcript or undefined if transcription fails or no audio is found.
*/
async function transcribeFirstAudio(params) {
	const { ctx, cfg } = params;
	const audioConfig = cfg.tools?.media?.audio;
	if (!audioConfig || audioConfig.enabled === false) return;
	const attachments = normalizeMediaAttachments(ctx);
	if (!attachments || attachments.length === 0) return;
	const firstAudio = attachments.find((att) => att && isAudioAttachment(att) && !att.alreadyTranscribed);
	if (!firstAudio) return;
	if (shouldLogVerbose()) logVerbose(`audio-preflight: transcribing attachment ${firstAudio.index} for mention check`);
	const providerRegistry = buildProviderRegistry(params.providers);
	const cache = createMediaAttachmentCache(attachments, { localPathRoots: resolveMediaAttachmentLocalRoots({
		cfg,
		ctx
	}) });
	try {
		const result = await runCapability({
			capability: "audio",
			cfg,
			ctx,
			attachments: cache,
			media: attachments,
			agentDir: params.agentDir,
			providerRegistry,
			config: audioConfig,
			activeModel: params.activeModel
		});
		if (!result || result.outputs.length === 0) return;
		const audioOutput = result.outputs.find((output) => output.kind === "audio.transcription");
		if (!audioOutput || !audioOutput.text) return;
		firstAudio.alreadyTranscribed = true;
		if (shouldLogVerbose()) logVerbose(`audio-preflight: transcribed ${audioOutput.text.length} chars from attachment ${firstAudio.index}`);
		return audioOutput.text;
	} catch (err) {
		if (shouldLogVerbose()) logVerbose(`audio-preflight: transcription failed: ${String(err)}`);
		return;
	} finally {
		await cache.cleanup();
	}
}

//#endregion
export { transcribeFirstAudio };
