// openclaw/dist/plugin-sdk/web-CVe_M9vU.js
import { a as WA_WEB_AUTH_DIR, g as webAuthExists, s as logWebSelfId, u as pickWebChannel } from "./accounts-D-nCmJLC.js";
import "./reply-K0V5vcRO.js";
import "./paths-DVWx7USN.js";
import "./github-copilot-token-Cg0YPPSu.js";
import "./plugins-B5m5N2-_.js";
import "./fetch-b2Zm8y7X.js";
import "./subsystem-B3gKmYuz.js";
import "./retry-qBs--rAQ.js";
import "./send-CPxMbC9v.js";
import "./config-D7yhe8WC.js";
import "./command-format-Ck33m20r.js";
import "./model-selection-Be0rkSSc.js";
import "./agent-scope-DeS5VuCp.js";
import "./manifest-registry-DNqjKxg2.js";
import "./dock-B-ZHp6hI.js";
import "./redact-4KGGYfvj.js";
import "./errors-1rgxcvGb.js";
import "./channel-activity-C0HQL-Ao.js";
import "./image-ops-CZuX7Z1C.js";
import "./ssrf-DswcqIq6.js";
import "./fetch-guard-DDo5yfUq.js";
import "./local-roots-g06aP0A4.js";
import "./ir-D8Kvtoe3.js";
import "./chunk-Ds3oYomW.js";
import "./message-channel-1SjHkh9S.js";
import "./bindings-DFv2Dwnp.js";
import "./markdown-tables-fj-uTU4P.js";
import "./render-C815sn1u.js";
import "./tables-D9ws6oLu.js";
import "./send-B1XX4o-3.js";
import "./tool-images-D8grjE7E.js";
import "./target-errors-B6sBXBAf.js";
import "./send-nEF74O5o.js";
import "./send-znTQy4eS.js";
import "./runner-CrEoy4qC.js";
import "./reply-prefix-Dznj6LTI.js";
import "./tokens-CcTm0O-e.js";
import "./skill-commands-CFDS7ojG.js";
import "./skills-Gt9vbHJR.js";
import "./chrome-B6OIe2xL.js";
import "./accounts-B6MN4-iI.js";
import "./accounts-DInAeEqh.js";
import "./deliver-DfGIcNwP.js";
import "./sessions-Kufe0AiW.js";
import "./paths-KtBBYvS4.js";
import "./diagnostic-AJZhyP1P.js";
import "./store-CXYl4NNL.js";
import "./pi-embedded-helpers-B_jcTbpF.js";
import { n as monitorWebInbox, t as monitorWebChannel } from "./channel-web-BifwIwTZ.js";
import "./thinking-DFs8hJTB.js";
import "./image-DvMwSfaa.js";
import "./pi-model-discovery-CLdAHc9I.js";
import "./api-key-rotation-DtqAmYo3.js";
import "./query-expansion-WDXVeP5-.js";
import "./diagnostic-session-state-xHsxBcGz.js";
import "./manager-LAaw4eYC.js";
import "./commands-registry-aYNO-rZL.js";
import "./send-DinoA1Sf.js";
import "./proxy-LN4R-K_y.js";
import "./resolve-route-DCVfVBNE.js";
import "./replies-9Mo9q0lI.js";
import "./outbound-attachment-YieXoD9L.js";
import { n as sendMessageWhatsApp } from "./outbound-C0QEbNER.js";
import { i as waitForWaConnection, t as createWaSocket } from "./session-DULidOZf.js";
import { t as loginWeb } from "./login-N7sAzl7G.js";

export { WA_WEB_AUTH_DIR, createWaSocket, logWebSelfId, loginWeb, monitorWebChannel, monitorWebInbox, pickWebChannel, sendMessageWhatsApp, waitForWaConnection, webAuthExists };
