// openclaw/dist/plugin-sdk/whatsapp-actions-D52B7D6e.js
import { r as resolveWhatsAppAccount } from "./accounts-D-nCmJLC.js";
import "./paths-DVWx7USN.js";
import "./github-copilot-token-Cg0YPPSu.js";
import "./plugins-B5m5N2-_.js";
import "./subsystem-B3gKmYuz.js";
import "./config-D7yhe8WC.js";
import "./command-format-Ck33m20r.js";
import "./model-selection-Be0rkSSc.js";
import "./agent-scope-DeS5VuCp.js";
import "./manifest-registry-DNqjKxg2.js";
import "./image-ops-CZuX7Z1C.js";
import "./ssrf-DswcqIq6.js";
import "./fetch-guard-DDo5yfUq.js";
import "./local-roots-g06aP0A4.js";
import "./ir-D8Kvtoe3.js";
import "./chunk-Ds3oYomW.js";
import "./message-channel-1SjHkh9S.js";
import "./bindings-DFv2Dwnp.js";
import "./markdown-tables-fj-uTU4P.js";
import "./render-C815sn1u.js";
import "./tables-D9ws6oLu.js";
import "./tool-images-D8grjE7E.js";
import { a as createActionGate, c as jsonResult, d as readReactionParams, i as ToolAuthorizationError, m as readStringParam } from "./target-errors-B6sBXBAf.js";
import { t as resolveWhatsAppOutboundTarget } from "./resolve-outbound-target-JSN_BQl_.js";
import { r as sendReactionWhatsApp } from "./outbound-C0QEbNER.js";

//#region src/agents/tools/whatsapp-target-auth.ts
function resolveAuthorizedWhatsAppOutboundTarget(params) {
	const account = resolveWhatsAppAccount({
		cfg: params.cfg,
		accountId: params.accountId
	});
	const resolution = resolveWhatsAppOutboundTarget({
		to: params.chatJid,
		allowFrom: account.allowFrom ?? [],
		mode: "implicit"
	});
	if (!resolution.ok) throw new ToolAuthorizationError(`WhatsApp ${params.actionLabel} blocked: chatJid "${params.chatJid}" is not in the configured allowFrom list for account "${account.accountId}".`);
	return {
		to: resolution.to,
		accountId: account.accountId
	};
}

//#endregion
//#region src/agents/tools/whatsapp-actions.ts
async function handleWhatsAppAction(params, cfg) {
	const action = readStringParam(params, "action", { required: true });
	const isActionEnabled = createActionGate(cfg.channels?.whatsapp?.actions);
	if (action === "react") {
		if (!isActionEnabled("reactions")) throw new Error("WhatsApp reactions are disabled.");
		const chatJid = readStringParam(params, "chatJid", { required: true });
		const messageId = readStringParam(params, "messageId", { required: true });
		const { emoji, remove, isEmpty } = readReactionParams(params, { removeErrorMessage: "Emoji is required to remove a WhatsApp reaction." });
		const participant = readStringParam(params, "participant");
		const accountId = readStringParam(params, "accountId");
		const fromMeRaw = params.fromMe;
		const fromMe = typeof fromMeRaw === "boolean" ? fromMeRaw : void 0;
		const resolved = resolveAuthorizedWhatsAppOutboundTarget({
			cfg,
			chatJid,
			accountId,
			actionLabel: "reaction"
		});
		const resolvedEmoji = remove ? "" : emoji;
		await sendReactionWhatsApp(resolved.to, messageId, resolvedEmoji, {
			verbose: false,
			fromMe,
			participant: participant ?? void 0,
			accountId: resolved.accountId
		});
		if (!remove && !isEmpty) return jsonResult({
			ok: true,
			added: emoji
		});
		return jsonResult({
			ok: true,
			removed: true
		});
	}
	throw new Error(`Unsupported WhatsApp action: ${action}`);
}

//#endregion
export { handleWhatsAppAction };
